test/unittest/tests/test_futex.cpp

```diff
@@ -0,0 +1,329 @@
+    UnexpectedTimeouts, /* More timeouts than expected. */
+
+    /* The following is a test implementation error. */
+    LostWakeup /* Waiter awoken without corresponding wake up signal. */
+};
+
+ostream &
+operator<<(ostream &out, const outcome &result)
+{
+#ifdef R_
+#error "R_ already defined"
+#endif
+
+#define R_(r)      \
+    case r:        \
+        out << #r; \
+        break
+
+    switch (result) {
+        R_(AsExpected);
+        R_(Error);
+        R_(UnexpectedTimeouts);
+        R_(LostWakeup);
+    }
+
+#undef R_
+    return out;
+}
+
+struct wake_signal {
+    WT_FUTEX_WAKE _type;
+    futex_word _value;
+
+    wake_signal(WT_FUTEX_WAKE t, futex_word v) : _type(t), _value(v) {}
+};
+
+struct wake_one : public wake_signal {
+    wake_one(futex_word val) : wake_signal(WT_FUTEX_WAKE_ONE, val) {}
+};
+
+struct wake_all : public wake_signal {
+    wake_all(futex_word val) : wake_signal(WT_FUTEX_WAKE_ALL, val) {}
+};
+
+bool
+is_wake_all(const wake_signal &wsig)
+{
+    return wsig._type == WT_FUTEX_WAKE_ALL;
+}
+
+} // namespace
+
+/*
+ * Reify for each test run: do NOT reuse.
+ */
+class futex_tester {
+public:
+    const chrono::duration<time_t, milli> INTER_WAKE_DELAY{1};
+
+    futex_word _futex;
+    vector<waiter> _waiters;
+    vector<thread> _threads;
+
+    void
+    create_waiters(size_t count, futex_word expected)
+    {
+        for (size_t i = 0; i < count; i++)
+            _waiters.push_back(waiter(expected));
+    }
+
+    void
+    start_waiters(futex_word expected, time_t timeout_usec)
+    {
+        futex_atomic_store(&_futex, expected);
+        for (auto &w : _waiters)
+            _threads.push_back(thread(&waiter::wait_on, &w, &_futex, timeout_usec));
+    }
+
+    void
+    delay_then_wake(time_t wake_start_delay_usec, const vector<wake_signal> &wake_signals)
+    {
+        chrono::duration<time_t, micro> WAKE_DELAY{wake_start_delay_usec};
+        this_thread::sleep_for(WAKE_DELAY);
+        for (auto &&sig : wake_signals) {
+            REQUIRE(__wt_futex_wake(&_futex, sig._type, sig._value) == 0);
+            this_thread::sleep_for(INTER_WAKE_DELAY);
+        }
+    }
+
+    void
+    wait_and_check(const vector<wake_signal> &wake_signals)
+    {
+        for (auto &&t : _threads)
+            t.join();
+
+        CAPTURE(_waiters);
+        CAPTURE(wake_signals);
+
+        auto result = inspect_waiters(wake_signals);
+        CAPTURE(result);
+        REQUIRE(result == outcome::AsExpected);
+    }
+
+    outcome
+    inspect_waiters(const vector<wake_signal> &wake_sigs)
+    {
+        /* Test validity check. */
+        REQUIRE(_waiters.size() >= wake_sigs.size());
+
+        auto wbeg = _waiters.cbegin();
+        auto wend = _waiters.cend();
+
+        /* Presence of any error other than timeout is failure. */
+        auto err_cnt = count_if(wbeg, wend, bind(&waiter::error, _1));
+        if (err_cnt > 0) {
```

Comment:
Nit: I think our style guide prefers no `{` `}` in this case. Same in a few more places below.

Reply:
removed x 3